src/array/cpu/spmat_op_impl.cc

```diff
@@ -0,0 +1,619 @@
+namespace impl {
+namespace {
+/*!
+ * \brief A hashmap that maps each ids in the given array to new ids starting from zero.
+ */
+template <typename IdType>
+class IdHashMap {
+ public:
+  // Construct the hashmap using the given id arrays.
+  // The id array could contain duplicates.
+  explicit IdHashMap(IdArray ids): filter_(kFilterSize, false) {
+    const IdType* ids_data = static_cast<IdType*>(ids->data);
+    const int64_t len = ids->shape[0];
+    IdType newid = 0;
+    for (int64_t i = 0; i < len; ++i) {
+      const IdType id = ids_data[i];
+      if (!Contains(id)) {
+        oldv2newv_[id] = newid++;
+        filter_[id & kFilterMask] = true;
+      }
+    }
+  }
+
+  // Return true if the given id is contained in this hashmap.
+  bool Contains(IdType id) const {
+    return filter_[id & kFilterMask] && oldv2newv_.count(id);
+  }
+
+  // Return the new id of the given id. If the given id is not contained
+  // in the hash map, returns the default_val instead.
+  IdType Map(IdType id, IdType default_val) const {
+    if (filter_[id & kFilterMask]) {
+      auto it = oldv2newv_.find(id);
+      return (it == oldv2newv_.end()) ? default_val : it->second;
+    } else {
+      return default_val;
+    }
+  }
+
+ private:
+  static constexpr int32_t kFilterMask = 0xFFFFFF;
+  static constexpr int32_t kFilterSize = kFilterMask + 1;
+  // This bitmap is used as a bloom filter to remove some lookups.
+  // Hashtable is very slow. Using bloom filter can significantly speed up lookups.
+  std::vector<bool> filter_;
+  // The hashmap from old vid to new vid
+  std::unordered_map<IdType, IdType> oldv2newv_;
+};
+
+struct PairHash {
+  template <class T1, class T2>
+  std::size_t operator() (const std::pair<T1, T2>& pair) const {
+    return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
+  }
+};
+
+template <typename DType>
+inline runtime::NDArray VecToNDArray(const std::vector<DType>& vec,
+    DLDataType dtype, DLContext ctx) {
+  const int64_t len = vec.size();
+  NDArray ret_arr = NDArray::Empty({len}, dtype, ctx);
+  DType* ptr = static_cast<DType*>(ret_arr->data);
+  std::copy(vec.begin(), vec.end(), ptr);
+  return ret_arr;
+}
+
+inline bool CSRHasData(CSRMatrix csr) {
+  return csr.data.defined();
+}
+
+inline bool COOHasData(COOMatrix csr) {
+  return csr.data.defined();
+}
+}  // namespace
+
+///////////////////////////// CSRIsNonZero /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+bool CSRIsNonZero(CSRMatrix csr, int64_t row, int64_t col) {
+  CHECK(row >= 0 && row < csr.num_rows) << "Invalid row index: " << row;
+  CHECK(col >= 0 && col < csr.num_cols) << "Invalid col index: " << col;
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  const IdType* indices_data = static_cast<IdType*>(csr.indices->data);
+  for (IdType i = indptr_data[row]; i < indptr_data[row + 1]; ++i) {
+    if (indices_data[i] == col) {
+      return true;
+    }
+  }
+  return false;
+}
```

Comment:
Seems we are missing the version for `runtime::NDArray CSRIsNonZero(CSRMatrix, runtime::NDArray row, runtime::NDArray col);`?

Reply:
You are right. That is `HasEdgesBetween(runtime::NDArray src, runtime::NDArray dst)`. I might forget to move that here.

Reply:
Done.